client/express-checkout/index.js

```diff
@@ -177,50 +188,104 @@ jQuery( ( $ ) => {
+
+				return options.displayItems
+					.filter( ( i ) => i.label === 'Shipping' )
+					.map( ( i ) => ( {
+						id: `rate-${ i.label }`,
+						amount: i.amount,
+						displayName: i.label,
+					} ) );
+			};
+
+			const shippingRates = getShippingRates();
+
+			// This is a bit of a hack, but we need some way to get the shipping information before rendering the button, and
+			// since we don't have any address information at this point it seems best to rely on what came with the cart response.
+			// Relying on what's provided in the cart response seems safest since it should always include a valid shipping
+			// rate if one is required and available.
+			// If no shipping rate is found we can't render the button so we just exit.
+			if ( options.requestShipping && ! shippingRates ) {
+				return;
+			}
+
 			const elements = api.getStripe().elements( {
 				mode: options?.mode ?? 'payment',
 				amount: options?.total,
 				currency: options?.currency,
+				paymentMethodCreation: 'manual',
 			} );
 
 			const eceButton = wcpayECE.createButton( elements, {
 				buttonType: {
-					googlePay: wcpayExpressCheckoutParams.button.type,
-					applePay: wcpayExpressCheckoutParams.button.type,
+					googlePay: getExpressCheckoutData( 'button' ).type,
+					applePay: getExpressCheckoutData( 'button' ).type,
 				},
 			} );
 
 			wcpayECE.showButton( eceButton );
 
-			wcpayECE.attachButtonEventListeners( eceButton );
-
 			eceButton.on( 'click', function ( event ) {
+				// TODO: handle cases where we need login confirmation.
+
+				if ( getExpressCheckoutData( 'is_product_page' ) ) {
+					wcpayECE.addToCart();
+				}
+
 				const clickOptions = {
-					business: {
-						name: 'Mikes Bikes',
-					},
-					lineItems: [
-						{ name: 'Bike', amount: 200 },
-						{ name: 'Helmet', amount: 300 },
-					],
-					shippingAddressRequired: true,
-					shippingRates: [
-						{
-							id: '1',
-							amount: 500,
-							displayName: 'Standard Shipping',
-						},
-						{
-							id: '2',
-							amount: 1000,
-							displayName: 'Expedited Shipping',
-						},
-					],
+					lineItems: options.displayItems.map( ( i ) => ( {
+						...i,
+						name: i.label,
+					} ) ),
+					emailRequired: true,
+					shippingAddressRequired: options.requestShipping,
+					phoneNumberRequired: options.requestPhone,
+					shippingRates,
 				};
+				wcpayECE.block();
 				event.resolve( clickOptions );
 			} );
 
+			eceButton.on( 'shippingaddresschange', ( event ) =>
+				shippingAddressChangeHandler( api, event, elements )
+			);
+
+			eceButton.on( 'shippingratechange', async ( event ) =>
+				shippingRateChangeHandler( api, event, elements )
+			);
+
+			eceButton.on( 'confirm', async ( event ) =>
+				onConfirmHandler(
+					api,
+					api.getStripe(),
+					elements,
+					wcpayECE.completePayment,
+					wcpayECE.abortPayment,
+					event
+				)
+			);
+
 			eceButton.on( 'cancel', () => {
+				wcpayECE.unblock();
 				wcpayECE.paymentAborted = true;
```

Comment:
I am not seeing where `paymentAborted` is being used for `express-checkout`. Will it be used, later? Or can it be removed?

Reply:
I don't know where it's used either. I didn't touch it because I was worried it'd break _something_. It might be used for tracks somewhere? I didn't look into it.

Reply:
It was copied from the PRB implementation. For PRBs, it's [used](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-request/index.js#L477) to reinitialize the button when the variation changes.

I'm going to remove it since we don't use it currently. If necessary, we could reintroduce it when working on [#8871](https://github.com/Automattic/woocommerce-payments/issues/8871) .

Reply:
done [766a675](https://github.com/Automattic/woocommerce-payments/pull/8937/commits/766a675a1eb1b0062d2788bc18611da905bb257d)